### ISSUE TITLE
com.hazelcast.jet.core.metrics.MetricsTest.customUnit_notUsed [HZ-997]

### ISSUE THREAD

**Issue by slavagithub** [hazelcast/hazelcast#19946]
_5.0.z_ (commit ce4ba622a7846f34def924fb13c5c3e613680523)

Failed on Windows, Oracle JDK8: [test report](http://jenkins.hazelcast.com/view/Official%20Builds/job/Hazelcast-5.maintenance-Windows-OracleJDK8/18/testReport/junit/com.hazelcast.jet.core.metrics/MetricsTest/customUnit_notUsed/)

<details><summary>Stacktrace:</summary>

```
java.lang.AssertionError: Expected measurements for metric 'sum', but there were none
	at org.junit.Assert.fail(Assert.java:89)
	at org.junit.Assert.assertTrue(Assert.java:42)
	at org.junit.Assert.assertFalse(Assert.java:65)
	at com.hazelcast.jet.core.metrics.JobMetricsChecker.assertAggregatedMetricValue(JobMetricsChecker.java:76)
	at com.hazelcast.jet.core.metrics.JobMetricsChecker.assertSummedMetricValue(JobMetricsChecker.java:54)
	at com.hazelcast.jet.core.metrics.MetricsTest.customUnit_notUsed(MetricsTest.java:164)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:115)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:107)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
```

</details>

<details><summary>Standard output:</summary>

```
0:01:00,933  INFO |customUnit| - [JetTestSupport] Thread-44565 - Terminating instanceFactory in JetTestSupport.@After
Finished Running Test: nonCooperativeProcessor in 0.713 seconds.
20:01:00,947 DEBUG |metricInFusedStages| - [JobCoordinationService] hz.determined_austin.cached.thread-4 - [127.0.0.1]:5701 [dev] [5.0.1-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
20:01:00,947  INFO |customUnit| - [LifecycleService] Thread-44573 - [127.0.0.1]:5701 [dev] [5.0.1-SNAPSHOT] [127.0.0.1]:5701 is SHUTTING_DOWN
20:01:00,948  WARN |customUnit| - [Node] Thread-44573 - [127.0.0.1]:5701 [dev] [5.0.1-SNAPSHOT] Terminating forcefully...
20:01:00,948  INFO |customUnit| - [Node] Thread-44573 - [127.0.0.1]:5701 [dev] [5.0.1-SNAPSHOT] Shutting down connection manager...
20:01:00,948  INFO |customUnit| - [Node] Thread-44573 - [127.0.0.1]:5701 [dev] [5.0.1-SNAPSHOT] Shutting down node engine...
```

</details>

**Comment by frant-hartm** [comment]
I was able to reproduce this with the following setting:

- in the test
```
        config.setProperty("hazelcast.jet.job.scan.period", "10");
```
- options:
``` 
-ea -javaagent:/home/frantisek/sources/rohypnol/core/target/rohypnol-core-0.2-SNAPSHOT-allinone.jar -Drohypnol.class.prefix=com.hazelcast -XX:+UnlockDiagnosticVMOptions -XX:+StressLCM -XX:+StressGCM
```
See https://github.com/jerrinot/rohypnol/ for the rohypnol library. cc @jerrinot 

Then just use the test repeat feature in intellij, every ~30th run fails.

I think that the following happens:
- when a job is finalized then jobResult (1) is created and after that master context is deleted (2)
- `JobCoordinationService.startJobIfNotStartedOrCompleted` called from `JobCoordinationService.doScanJobs` checks this in the same order - check job result (3) and putIfAbsent for master context (4)

So this order of actions is possible:
3, 1, 2, 4
at this point, we have a master context for a job that already has a result and its original master context has already been removed. This is then handled by `completeMasterContextIfJobAlreadyCompleted` (5)

The problem is if there is an action using `JobCoordinationService#callWithJob` between 4 and 5 - in the case of this test failure it returns empty metrics because it collects metrics using `MasterJobContext#collectMetrics` and doesn't use already stored metrics (it actually just returns empty list because the re-created MC has job state `NOT_RUNNING` and empty list is the initial value for jobMetrics).

The following test fails for another operation using `callWithJob` - `job.getStatus()`:
```
    @Test
    public void job_status() {
        pipeline.readFrom(TestSources.items(0L, 1L, 2L, 3L, 4L))
                .mapStateful(LongAccumulator::new, (acc, i) -> {
                    acc.add(i);
                    Metrics.metric("sum", Unit.COUNT);
                    return acc.get();
                })
                .writeTo(Sinks.noop());

        Job job = runPipeline(pipeline.toDag()); // there is job.join() inside
        assertThat(job.getStatus()).isEqualTo(JobStatus.COMPLETED);
    }


Expected :COMPLETED
Actual   :NOT_RUNNING
```
So this is a wider issue.

A naive fix would be to put the `MasterContext` removal under synchronization as well and move the check job result (3) into the same synchronized block as `putIfAbsent` in `startJobIfNotStartedOrCompleted` - https://github.com/hazelcast/hazelcast/blob/12691796a7de6b9c4bbbcc02fe013fb34a31f84f/hazelcast/src/main/java/com/hazelcast/jet/impl/JobCoordinationService.java#L1080-L1092

@viliam-durina feedback welcome :-)
